amqp/src/AmqpClient.cpp

```diff
@@ -19,160 +19,287 @@
-    m_senderReady.wait(l);
-    logDebug("sender ready on {}", msg.to());
+    logError("Transport error: {}", transport.error().what());
+    m_communicationState = ComState::COM_STATE_LOST;
+  }
 
-    m_sender.work_queue().add([=]() {
-      auto tracker = m_sender.send(msg);
-      logDebug("Msg sent {}", proton::to_string(tracker.state()));
-      m_sender.connection().close();
-    });
+  void AmqpClient::resetPromise()
+  {
+    logDebug("Reset all promise");
+    m_connectPromise = std::promise<fty::messagebus::ComState>();
+    m_connectFuture = m_connectPromise.get_future();
+    m_promiseSender = std::promise<void>();
+    m_promiseReceiver = std::promise<void>();
+  }
+
+  ComState AmqpClient::connected()
+  {
+    if ((m_communicationState == ComState::COM_STATE_UNKNOWN) || (m_communicationState == ComState::COM_STATE_LOST))
+    {
+      if (m_connectFuture.wait_for(TIMEOUT) != std::future_status::timeout)
+      {
+        try
+        {
+          m_communicationState = m_connectFuture.get();
+        }
+        catch (const std::future_error& e)
+        {
+          logError("Caught future error {}", e.what());
+        }
+      }
+      else
+      {
+        m_communicationState = ComState::COM_STATE_CONNECT_FAILED;
+      }
+    }
+    return m_communicationState;
+  }
+
+  DeliveryState AmqpClient::send(const proton::message& msg)
+  {
+    auto deliveryState = DeliveryState::DELIVERY_STATE_REJECTED;
+    if (connected() == ComState::COM_STATE_OK)
+    {
+      m_promiseSender = std::promise<void>();
+      logDebug("Sending message to {} ...", msg.to());
+      m_message.clear();
+      m_message = msg;
+
+      m_connection.work_queue().add([=]() {
+        m_connection.open_sender(msg.to());
+      });
+
+      // Wait the to know if the message has been sent or not
+      if (m_promiseSender.get_future().wait_for(TIMEOUT) != std::future_status::timeout)
+      {
+        deliveryState = DeliveryState::DELIVERY_STATE_ACCEPTED;
+      }
+    }
+    return deliveryState;
+  }
+
+  DeliveryState AmqpClient::receive(const Address& address, const std::string& filter, MessageListener messageListener)
+  {
+    auto deliveryState = DeliveryState::DELIVERY_STATE_REJECTED;
+    if (connected() == ComState::COM_STATE_OK)
+    {
+      logDebug("Set receiver to wait message(s) from {} ...", address);
+      m_promiseReceiver = std::promise<void>();
+
+      auto futureReceiver = m_promiseReceiver.get_future();
+      (!filter.empty())? setSubscriptions(filter, messageListener) : setSubscriptions(address, messageListener);
+
+      m_connection.work_queue().add([=]() {
+        m_connection.open_receiver(address, {});
+      });
+
+      if (futureReceiver.wait_for(TIMEOUT) != std::future_status::timeout)
+      {
+        deliveryState = DeliveryState::DELIVERY_STATE_ACCEPTED;
+      }
+    }
+
+    return deliveryState;
   }
 
   bool AmqpClient::tryConsumeMessageFor(std::shared_ptr<proton::message> resp, int timeout)
```

Comment:
fir readability, here timeout shall be typed as chrono::seconds....
or at least named timeout_sec....